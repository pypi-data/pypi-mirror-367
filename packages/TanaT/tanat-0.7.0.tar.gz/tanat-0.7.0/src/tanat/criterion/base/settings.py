#!/usr/bin/env python3
"""
Base class for criterion.
"""

import dataclasses

from .enum import CriterionLevel


@dataclasses.dataclass
class Criterion:
    """
    Base class for criterion.
    """

    _REGISTER_NAME = None

    def get_compatibility_levels(self, max_level=CriterionLevel.TRAJECTORY):
        """List the compatibility levels of the criterion.
        Compatibility levels are marked with decorators.

        Args:
            max_level (CriterionLevel, optional): The maximum level to consider.
                Defaults to Trajectory. Entity < Sequence < Trajectory

        Returns a list of CriterionLevel representing the compatibility levels of the
        criterion. The levels are "sequence","entity" or "trajectory".
        """
        if isinstance(max_level, str):
            max_level = CriterionLevel.from_str(max_level)

        compatibility_levels = getattr(self, "COMPATIBILITY_LEVELS", set())
        return [level for level in compatibility_levels if level <= max_level]
