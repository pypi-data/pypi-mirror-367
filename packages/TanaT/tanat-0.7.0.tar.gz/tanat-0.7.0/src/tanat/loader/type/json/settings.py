#!/usr/bin/env python3
"""
Loader for JSON files.
"""

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer
from pypassist.fallback.typing import Dict


@viewer
@dataclass
class JSONLoaderSettings:
    """
    Settings for JSON loader

    Attributes:
        filepath: Path to the JSON file to load
        encoding: File encoding (utf-8 by default)
        json_load_kwargs: Additional arguments to pass to json.load
    """

    filepath: str
    encoding: str = "utf-8"
    json_load_kwargs: Dict = None

    def __post_init__(self):
        self.json_load_kwargs = self.json_load_kwargs or {}
