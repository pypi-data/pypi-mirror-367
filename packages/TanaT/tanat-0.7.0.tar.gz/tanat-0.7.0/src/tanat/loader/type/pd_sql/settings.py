#!/usr/bin/env python3
"""
Dataframe Loader Settings for SQL databases.
"""

from typing import Optional

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer
from pypassist.fallback.typing import Dict


@viewer
@dataclass
class PandasSQLLoaderSettings:
    """
    Settings for SQL loader

    Attributes:
        conn: SQLAlchemy connection string (e.g., 'sqlite:///database.db')
        query: SQL query to execute
        pd_read_sql_kwargs: Additional arguments to pass to pd.read_sql
    """

    conn: str
    query: str
    pd_read_sql_kwargs: Optional[Dict] = None

    def __post_init__(self):
        self.pd_read_sql_kwargs = self.pd_read_sql_kwargs or {}
