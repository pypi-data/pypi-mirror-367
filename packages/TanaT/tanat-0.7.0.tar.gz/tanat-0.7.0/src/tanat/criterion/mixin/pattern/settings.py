#!/usr/bin/env python3
"""
Pattern criterion.
"""

from typing import Union

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer
from pypassist.fallback.typing import Dict, List

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible
from .enum import PatternOperator


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class PatternCriterion(Criterion):
    """
    Flexible pattern-based filtering for sequences and entities.

    Enables complex pattern matching across data columns with advanced
    configuration options.

    Args:
        pattern (Dict[str, Union[str, List[str]]]):
            Column patterns to match. Supports:
            - Plain strings
            - Lists of strings
            - Regex patterns (prefix with 'regex:')

        contains (bool, optional):
            - True: Include sequences matching pattern
            - False: Exclude sequences matching pattern
            Defaults to True.

        case_sensitive (bool, optional):
            Whether matching is case-sensitive. Defaults to True.

        operator (PatternOperator, optional):
            Logical operator between patterns:
            - AND: All patterns must match
            - OR: At least one pattern matches
            Defaults to AND.

    Example:
        # Find sequences with emergency events
        PatternCriterion(pattern={'event_type': 'EMERGENCY'})
    """

    pattern: Dict[str, Union[str, List[str]]]
    contains: bool = True
    case_sensitive: bool = True
    operator: PatternOperator = PatternOperator.AND

    _REGISTER_NAME = "pattern"
