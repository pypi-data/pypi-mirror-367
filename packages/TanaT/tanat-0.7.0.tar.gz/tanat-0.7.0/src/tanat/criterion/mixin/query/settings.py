#!/usr/bin/env python3
"""
Settings for query criterion.
"""

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class QueryCriterion(Criterion):
    """
    Pandas-based filtering criterion for sequences and entities.

    Allows complex data filtering using pandas query syntax. Compatible with
    sequence and entity levels, enabling flexible data selection based on
    column conditions.

    Args:
        query (str): Pandas-style query string to filter data.
            Supports complex boolean conditions across columns.

    Example:
        # Filter sequences where age > 50 and gender is 'M'
        criterion = QueryCriterion(query="age > 50 and gender == 'M'")
    """

    query: str

    _REGISTER_NAME = "query"
