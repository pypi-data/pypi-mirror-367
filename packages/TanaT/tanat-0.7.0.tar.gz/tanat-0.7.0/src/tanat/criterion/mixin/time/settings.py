#!/usr/bin/env python3
"""
Settings for time-based filtering.
"""

from typing import Optional, Union
from datetime import datetime

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class TimeCriterion(Criterion):
    """
    Flexible time-based filtering for sequences and entities.

    Enables precise temporal filtering with multiple time constraints and
    different containment modes depending on the filtering level.

    Args:
        start_before (datetime, optional):
            Filter sequences or entities starting before this timestamp.
        start_after (datetime, optional):
            Filter sequences or entities starting after this timestamp.
        end_before (datetime, optional):
            Filter sequences or entities ending before this timestamp.
        end_after (datetime, optional):
            Filter sequences or entities ending after this timestamp.
        duration_within (bool, optional):
            For entity-level filtering on state/interval entities:
            - True: Entity must be entirely contained within the time range
            - False: Entity can partially overlap with the time range
            Defaults to False. Has no effect on event entities.
        sequence_within (bool, optional):
            For sequence-level filtering (event, interval, state):
            - True: Entire sequence must be contained within the time range
            - False: Sequence can partially overlap with the time range
            Defaults to False.
        date_format (str, optional):
            Custom datetime parsing format for string inputs.

    Examples:
        # Filter entities with partial overlap (default behavior)
        TimeCriterion(
            start_after=datetime.now(),
            end_before=datetime.now() + timedelta(days=90)
        )

        # Only keep state/interval entities fully contained in time range
        TimeCriterion(
            start_after=datetime(2024, 1, 1),
            end_before=datetime(2024, 12, 31),
            duration_within=True
        )

        # Only keep sequences entirely within time range
        TimeCriterion(
            start_after=datetime(2024, 6, 1),
            end_before=datetime(2024, 6, 30),
            sequence_within=True
        )

        # Combine both strict modes
        TimeCriterion(
            start_after=datetime(2024, 1, 1),
            end_before=datetime(2024, 12, 31),
            duration_within=True,
            sequence_within=True
        )
    """

    start_before: Optional[Union[datetime, str, int]] = None
    start_after: Optional[Union[datetime, str, int]] = None
    end_before: Optional[Union[datetime, str, int]] = None
    end_after: Optional[Union[datetime, str, int]] = None
    duration_within: bool = False
    sequence_within: bool = False
    date_format: Optional[str] = None
    # TODO: add settings for duration

    _REGISTER_NAME = "time"
