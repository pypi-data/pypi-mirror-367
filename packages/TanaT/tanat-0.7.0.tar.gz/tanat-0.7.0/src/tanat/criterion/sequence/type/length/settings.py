#!/usr/bin/env python3
"""
Length criterion.
"""

from typing import Optional

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ....base.settings import Criterion
from ....decorator.sequence import sequence_compatible
from .exception import ContradictoryLengthCriterionError


@viewer
@sequence_compatible
@dataclass
class LengthCriterion(Criterion):
    """
    Filter sequences based on their length.

    Provides flexible length-based filtering with multiple comparison
    operators for sequence selection.

    Args:
        gt (int, optional):
            Keep sequences with length strictly greater than this value.
        ge (int, optional):
            Keep sequences with length greater than or equal to this value.
        lt (int, optional):
            Keep sequences with length strictly less than this value.
        le (int, optional):
            Keep sequences with length less than or equal to this value.

    Example:
        # Filter sequences with 10 or fewer events
        LengthCriterion(le=10)
    """

    gt: Optional[int] = None  # greater than
    ge: Optional[int] = None  # greater than or equal to
    lt: Optional[int] = None  # less than
    le: Optional[int] = None  # less than or equal to

    _REGISTER_NAME = "length"

    def __post_init__(self):
        if self.gt is not None and self.ge is not None:
            raise ContradictoryLengthCriterionError(
                "Cannot use 'gt' and 'ge' operators simultaneously"
            )
        if self.lt is not None and self.le is not None:
            raise ContradictoryLengthCriterionError(
                "Cannot use 'lt' and 'le' operators simultaneously"
            )

        lower = self.gt if self.gt is not None else self.ge
        upper = self.lt if self.lt is not None else self.le
        if lower is not None and upper is not None and lower >= upper:
            raise ContradictoryLengthCriterionError(
                f"lower bound ({lower}) must be strictly less than upper bound ({upper})."
            )
