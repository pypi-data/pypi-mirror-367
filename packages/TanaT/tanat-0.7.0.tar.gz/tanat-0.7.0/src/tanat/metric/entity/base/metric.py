#!/usr/bin/env python3
"""
Base class for registable event metrics.
"""

from abc import abstractmethod

from .exception import UnregisteredEntityMetricTypeError
from .picklable import _EntityMetricPicklableComputer
from ....sequence.base.entity import Entity
from ...base import Metric


class EntityMetric(Metric):
    """
    Base class for event metrics.
    """

    _REGISTER = {}

    @abstractmethod
    def __call__(self, ent_a, ent_b, **kwargs):
        """
        Compute the metric for a specific pair of entities.

        Args:
            ent_a (Entity):
                First sequence.

            ent_b (Entity):
                Second sequence.

            kwargs:
                Optional arguments to override specific settings.

        Returns:
            float: The metric value for the entity pair.
        """

    def compute_sequence(self, seq_a, seq_b, **kwargs):
        """
        Compute the EntityMetric between two sequences.

        Args:
            seq_a (Sequence):
                First sequence.

            seq_b (Sequence):
                Second sequence.

            kwargs:
                Optional arguments to override specific settings.

        Returns:
            np.ndarray: An array of entity metric values.
        """
        self._validate_sequences(seq_a=seq_a, seq_b=seq_b)
        with self.with_tmp_settings(**kwargs):  # temporarily override settings
            data_a = seq_a.entities()
            data_b = seq_b.entities()
            yield from map(lambda ent_pair: self(*ent_pair), zip(data_a, data_b))

    def picklable_sequence_computer(self, **kwargs):
        """
        Return a picklable computer for parallel sequence metric computation.

        Returns a callable object that encapsulates the core computation logic
        (_compute_sequence_array) and settings in a way that can be serialized
        for parallel processing with multiprocessing.Pool.

        Args:
            kwargs:
                Optional arguments to override specific settings.

        Returns:
            _EntityMetricPicklableComputer: Picklable wrapper for sequence computation

        Note:
            This is the recommended way to use entity metrics in parallel computations,
            as it properly handles serialization of the computation logic and settings.
        """
        return _EntityMetricPicklableComputer.from_entity_metric(self, **kwargs)

    @staticmethod
    @abstractmethod
    def _compute_sequence_array(arrays_a, arrays_b, settings):
        """
        Compute EntityMetric between two sequences of arrays.

        Args:
            arrays_a: First sequence array
            arrays_b: Second sequence array
            settings: Configuration settings for the computation

        Returns:
            Array of distances (elementwise) or distance matrix (allpairs)
        """

    def _validate_entities(self, **entities):
        """
        Validate multiple sequences, ensuring they are of the correct type.

        Args:
            entities:
                Dictionary of entities to validate.

        Raises:
            ValueError: If any entity is invalid.
        """
        for key, entity in entities.items():
            if not self._is_valid_entity(entity):
                raise ValueError(
                    f"Invalid sequence '{key}'. Expected Entity, got {type(entity)}."
                )

    def _is_valid_entity(self, entity):
        """
        Check if a given sequence is valid.

        Args:
            entity:
                The entity to validate.

        Returns:
            bool: True if valid, False otherwise.
        """
        return isinstance(entity, Entity)

    @classmethod
    def _unregistered_metric_error(cls, mtype, err):
        """Raise an error for an unregistered entity metric with a custom message."""
        registered = cls.list_registered()
        raise UnregisteredEntityMetricTypeError(
            f"Unknown entity metric: '{mtype}'. " f"Available metrics: {registered}"
        ) from err
