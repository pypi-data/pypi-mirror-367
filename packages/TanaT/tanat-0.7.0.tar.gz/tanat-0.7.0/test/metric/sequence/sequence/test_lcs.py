#!/usr/bin/env python3
"""Test LCSSequenceMetric call method."""

import pytest

from tanat.metric.sequence.type.lcs.metric import (
    LCSSequenceMetric,
)
from tanat.metric.sequence.type.lcs.settings import (
    LCSSequenceMetricSettings,
)


class TestLCSSequenceMetric:
    """
    Test LCSSequenceMetric call method.
    """

    @pytest.mark.parametrize("pool_type", ["event", "state", "interval"])
    def test_call_with_snapshot(self, sequence_pools, pool_type, snapshot):
        """
        Ensure call method gives expected results for LCSSequenceMetric.
        Snapshot is used to check the actual content.
        """
        pool = sequence_pools[pool_type]
        # Initialize LCSSequenceMetric with default settings
        metric = LCSSequenceMetric(settings=LCSSequenceMetricSettings())

        ## -- sequence to compare
        seq_a = pool[2]
        seq_b = pool[3]

        # Call with SequenceMetric object
        value = metric(seq_a, seq_b)
        snapshot.assert_match(value)
