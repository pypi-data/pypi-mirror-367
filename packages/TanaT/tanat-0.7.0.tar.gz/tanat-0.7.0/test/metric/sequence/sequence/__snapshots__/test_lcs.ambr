# serializer version: 1
# name: TestLCSSequenceMetric.test_call_with_snapshot[event]
  2.0
# ---
# name: TestLCSSequenceMetric.test_call_with_snapshot[interval]
  0.0
# ---
# name: TestLCSSequenceMetric.test_call_with_snapshot[state]
  0.0
# ---
